firedrake/functionspaceimpl.py

```diff
@@ -0,0 +1,736 @@
+            # First dimension of the value_shape is the VectorElement
+            # shape.
+            self.shape = element.value_shape()[:1]
+        else:
+            self.shape = ()
+        self._ufl_element = element
+        self._shared_data = sdata
+        self._mesh = mesh
+
+        self.rank = len(self.shape)
+        """The rank of this :class:`FunctionSpace`.  Spaces where the
+        element is scalar-valued (or intrinsically vector-valued) have
+        rank zero.  Spaces built on :class:`~ufl.classes.VectorElement` or
+        :class:`~ufl.classes.TensorElement` instances have rank equivalent to
+        the number of components of their
+        :meth:`~ufl.classes.FiniteElementBase.value_shape`."""
+        self.dim = numpy.prod(self.shape, dtype=int)
+        """The total number of degrees of freedom at each function
+        space node."""
+        self.name = name
+        """The (optional) descriptive name for this space."""
+        self.node_set = sdata.node_set
+        """A :class:`pyop2.Set` representing the function space nodes."""
+        self.dof_dset = op2.DataSet(self.node_set, self.shape or 1,
+                                    name="%s_nodes_dset" % self.name)
+        """A :class:`pyop2.DataSet` representing the function space
+        degrees of freedom."""
+        self.fiat_element = fiat_element
+        self.extruded = sdata.extruded
+        self.offset = sdata.offset
+        self.bt_masks = sdata.bt_masks
+
+    # These properties are overridden in ProxyFunctionSpaces, but are
+    # provided by FunctionSpace so that we don't have to special case.
+    index = None
+    """The position of this space in its parent
+    :class:`MixedFunctionSpace`, or ``None``."""
+
+    parent = None
+    """The parent space if this space was extracted from one, or ``None``."""
+
+    component = None
+    """The component of this space in its parent VectorElement space, or
+    ``None``."""
+
+    def __eq__(self, other):
+        try:
+            # FIXME: Think harder about equality
+            return self.mesh() is other.mesh() and \
+                self.dof_dset is other.dof_dset and \
+                self.ufl_element() == other.ufl_element()
+        except AttributeError:
+            return False
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    @utils.cached_property
+    def _dm(self):
+        """A PETSc DM describing the data layout for this FunctionSpace."""
+        dm = PETSc.DMShell().create()
+        dm.setAttr('__fs__', weakref.ref(self))
+        dm.setPointSF(self.mesh()._plex.getPointSF())
+        dm.setDefaultSection(self._shared_data.global_numbering)
+        # FIXME: The Vec, but not the DM (since it holds a reference
+        # to the function space) could be shared between
+        # functionspaces with the same dof_dset.
+        with self.make_dat().vec_ro as v:
+            dm.setGlobalVector(v.duplicate())
```

Comment:
I claim `FunctionSpaces` are now lightweight, however this is slightly bad.  I need one dm per `FunctionSpace` instance (that's OK, it's not big).  The `pointSF` is shared and the default section is shared.  However, the `Vec` is not shared.  So each instance of a `FunctionSpace` that accesses the `_dm` property makes a new `Vec` (i.e. a new `Dat`).  The Vec can be shared with everything that has the same number of `nodes_per_entity` which gives the global numbering, and also the same dofs per node (i.e. the function space shape).  Miklos' suggestion of caching individual bits would fix this problem I think.